tests/subsys/rtio/rtio_api/src/test_rtio_api.c

```diff
@@ -333,6 +337,136 @@ ZTEST(rtio_api, test_rtio_simple_mempool)
 #endif
 }
 
+static void test_rtio_simple_cancel_(struct rtio *r, int i)
+{
+	struct rtio_sqe sqe[SQE_POOL_SIZE];
+	struct rtio_cqe cqe;
+	struct rtio_sqe *handle[1];
+
+	rtio_sqe_prep_nop(sqe, (struct rtio_iodev *)&iodev_test_simple, NULL);
+	rtio_sqe_copy_in_get_handles(r, sqe, handle, 1);
+	rtio_sqe_cancel(handle[0]);
+	TC_PRINT("Submitting 1 to RTIO\n");
+	rtio_submit(r, 0);
+
+	/* Check that we don't get a CQE */
+	zassert_equal(0, rtio_cqe_copy_out(r, &cqe, 1, K_MSEC(15)));
+
+	/* Check that the SQE pool is empty by filling it all the way */
+	for (int i = 0; i < SQE_POOL_SIZE; ++i) {
+		rtio_sqe_prep_nop(&sqe[i], (struct rtio_iodev *)&iodev_test_simple, NULL);
+	}
+	zassert_ok(rtio_sqe_copy_in(r, sqe, SQE_POOL_SIZE));
+
+	/* Since there's no good way to just reset the RTIO context, wait for the nops to finish */
+	rtio_submit(r, SQE_POOL_SIZE);
+	for (int i = 0; i < SQE_POOL_SIZE; ++i) {
+		zassert_equal(1, rtio_cqe_copy_out(r, &cqe, 1, K_FOREVER));
+	}
+}
+
+static void rtio_simple_cancel_test(void *a, void *b, void *c)
+{
+	ARG_UNUSED(a);
+	ARG_UNUSED(b);
+	ARG_UNUSED(c);
+
+	TC_PRINT("rtio simple cancel\n");
+	for (int i = 0; i < TEST_REPEATS; i++) {
+		test_rtio_simple_cancel_(&r_simple, i);
+	}
+}
+
+ZTEST(rtio_api, test_rtio_simple_cancel)
+{
+	rtio_iodev_test_init(&iodev_test_simple);
+#ifdef CONFIG_USERSPACE
+	k_mem_domain_add_thread(&rtio_domain, k_current_get());
+	rtio_access_grant(&r_simple, k_current_get());
+	k_object_access_grant(&iodev_test_simple, k_current_get());
+	k_thread_user_mode_enter(rtio_simple_cancel_test, NULL, NULL, NULL);
+#else
+	rtio_simple_cancel_test(NULL, NULL, NULL);
+#endif
+}
```

Comment:
Could we get a test that also hits the [Chained Blocking  Requests](https://docs.zephyrproject.org/latest/services/rtio/index.html#chained-blocking-requests) scenario?  Is it possible to have orphaned SQEs if somehow the write is canceled while the read isn't in linked example above? or do all items in the chain get canceled in this scenario?

Reply:
Added tests for transactions and chains.